cmd/runtimetest/main.go

```diff
@@ -278,63 +354,148 @@ func validateRlimits(spec *rspec.Spec, t *tap.T) error {
 			return err
 		}
 
-		if rlimit.Cur != r.Soft {
-			return specerror.NewError(specerror.PosixProcRlimitsSoftMatchCur, fmt.Errorf("%v rlimit soft expected: %v, actual: %v", r.Type, r.Soft, rlimit.Cur), rspec.Version)
+		rfcError, err := c.Ok(rlimit.Cur == r.Soft, specerror.PosixProcRlimitsSoftMatchCur, spec.Version, fmt.Sprintf("has expected soft %v", r.Type))
+		if err != nil {
+			return err
 		}
-		if rlimit.Max != r.Hard {
-			return specerror.NewError(specerror.PosixProcRlimitsHardMatchMax, fmt.Errorf("%v rlimit hard expected: %v, actual: %v", r.Type, r.Hard, rlimit.Max), rspec.Version)
+		c.harness.YAML(map[string]interface{}{
+			"level":     rfcError.Level.String(),
+			"reference": rfcError.Reference,
+			"type":      r.Type,
+			"expected":  r.Soft,
+			"actual":    rlimit.Cur,
+		})
+
+		rfcError, err = c.Ok(rlimit.Max == r.Hard, specerror.PosixProcRlimitsHardMatchMax, spec.Version, fmt.Sprintf("has expected hard %v", r.Type))
+		if err != nil {
+			return err
 		}
+		c.harness.YAML(map[string]interface{}{
+			"level":     rfcError.Level.String(),
+			"reference": rfcError.Reference,
+			"type":      r.Type,
+			"expected":  r.Hard,
+			"actual":    rlimit.Max,
+		})
 	}
 	return nil
 }
 
-func validateSysctls(spec *rspec.Spec, t *tap.T) error {
-	if spec.Linux == nil {
+func (c *complianceTester) validateSysctls(spec *rspec.Spec) error {
+	if spec.Linux == nil || spec.Linux.Sysctl == nil {
+		c.harness.Skip(1, "linux.sysctl not set")
 		return nil
 	}
+
 	for k, v := range spec.Linux.Sysctl {
 		keyPath := filepath.Join("/proc/sys", strings.Replace(k, ".", "/", -1))
 		vBytes, err := ioutil.ReadFile(keyPath)
 		if err != nil {
 			return err
 		}
 		value := strings.TrimSpace(string(bytes.Trim(vBytes, "\x00")))
-		if value != v {
-			return fmt.Errorf("Sysctl %v value expected: %v, actual: %v", k, v, value)
-		}
+		c.harness.Ok(value == v, fmt.Sprintf("has expected sysctl %v", k))
+		c.harness.YAML(map[string]string{
+			"sysctl":   k,
+			"expected": v,
+			"actual":   value,
+		})
 	}
 	return nil
 }
 
-func testWriteAccess(path string) error {
-	tmpfile, err := ioutil.TempFile(path, "Test")
+func testReadAccess(path string) (readable bool, err error) {
+	fi, err := os.Stat(path)
 	if err != nil {
-		return err
+		return false, err
 	}
+	if fi.Mode()&os.ModeType == 0 {
+		return testFileReadAccess(path)
+	}
+	return false, fmt.Errorf("cannot test read access for %q (mode %d)", path, fi.Mode())
+}
+
+func testFileReadAccess(path string) (readable bool, err error) {
+	f, err := os.Open(path)
+	if err != nil {
+		return false, nil
+	}
+	defer f.Close()
+	b := make([]byte, 1)
+	_, err = f.Read(b)
+	if err == nil {
+		return true, nil
+	} else if err == io.EOF {
+		return false, nil
```

Comment:
I find this surprising to return "non-readable" when the file might well be readable but just empty.

Reply:
> I find this surprising to return "non-readable" when the file might well be readable but just empty.

Yeah, but I'm not sure how to detect readable-but-empty (and what I have here is what we've [been using][1]).  Do you have ideas?  The lack of support for that case makes this brittle, but I'm ok with it if we update our tests (e.g. [here][2]) to use non-empty test files.  `runtimetest` is for internal testing only (not part of our public API), so a few cludges to work around the `runtimetest` limitation are unfortunate but survivable.

[1]: https://github.com/opencontainers/runtime-tools/blob/fcd5b23b0f05aa34cb922fe39c6ba79e6f46fe39/cmd/runtimetest/main.go#L534-L543
[2]: https://github.com/opencontainers/runtime-tools/blob/fcd5b23b0f05aa34cb922fe39c6ba79e6f46fe39/validation/linux_readonly_paths.go#L14-L15

Reply:
Actually, our current tests use an empty file and [treat an immediate EOF as success][1].  So I am changing the logic.  But the old logic doesn't make much sense to me.  What if the runtime implemented `linux.readonlyPaths` by binding `/dev/null` over the listed paths?  That would seem to satisfy [the spec][2] (which is pretty weak on this anyway), while still resulting in immediate EOFs.

[1]: https://github.com/opencontainers/runtime-tools/blob/fcd5b23b0f05aa34cb922fe39c6ba79e6f46fe39/cmd/runtimetest/main.go#L540-L542
[2]: https://github.com/opencontainers/runtime-spec/blob/v1.0.1/config-linux.md#readonly-paths

Reply:
Maybe just add a comment to explain the situation:
```
	} else if err == io.EOF {
                // linux_readonly_paths.go only tests non-empty files. So if we find
                // an empty file, the runtime did not apply the ReadonlyPaths correctly
		return false, nil
```

I agree about updating the test you mention :)